### PR TITLE
When reading files, check if the eof flag is set before throwing.

### PR DESCRIPTION
**Pull request by xyz32**
#3039

### CODE REVIEW

File: lib/system.nim [hunk 1]
```diff
@@ -2594,7 +2594,10 @@ when not defined(JS): #and not defined(nimscript):
       ## Closes the file.
 
     proc endOfFile*(f: File): bool {.tags: [], benign.}
-      ## Returns true iff `f` is at the end.
+      ## Returns true if `f` is at the end.
```
Comment (dom96) [inline lib/system.nim]: "iff" stands for "if and only if" :)

Comment (reactormonk) [inline lib/system.nim]: Don't worry, I was in the exact same situation a few years ago :-)

File: lib/system/sysio.nim [hunk 2]
```diff
@@ -37,6 +37,8 @@ proc fread(buf: pointer, size, n: int, f: File): int {.
 proc fseek(f: File, offset: clong, whence: int): int {.
   importc: "fseek", header: "<stdio.h>", tags: [].}
 proc ftell(f: File): int {.importc: "ftell", header: "<stdio.h>", tags: [].}
+proc ferror(f: File): int {.importc: "ferror", header: "<stdio.h>", tags: [].}
+proc feof(f: File): int {.importc: "feof", header: "<stdio.h>", tags: [].}
```
Comment (dom96) [inline lib/system/sysio.nim]: is this actually used?

**Comment by dom96** [comment]
Other than that I think this is fine. But I will leave it to @Araq to review as well.

File: lib/system/sysio.nim [hunk 3]
```diff
@@ -147,9 +149,17 @@ proc rawFileSize(file: File): int =
 proc readAllFile(file: File, len: int): string =
   # We acquire the filesize beforehand and hope it doesn't change.
   # Speeds things up.
-  result = newString(int(len))
-  if readBuffer(file, addr(result[0]), int(len)) != len:
+  result = newString(len)
+  let bytes = readBuffer(file, addr(result[0]), len)
+  if endOfFile(file):
+    if bytes < len:
+      result = substr(result, 0 , bytes - 1)
```
Comment (Araq) [inline lib/system/sysio.nim]: Use 'setLen' instead of 'substr'.

File: lib/system/sysio.nim [hunk 4]
```diff
@@ -147,9 +149,17 @@ proc rawFileSize(file: File): int =
...
+  elif fileError(file):
     raiseEIO("error while reading from file")
+  else:
+    # We red all the bytes but did not reach the EOF
```
Comment (Araq) [inline lib/system/sysio.nim]: 'We read ...'

**Comment by Araq** [comment]
This is not how to fix it. Instead if ftell() returns -1 it should go into a "reading fixed size buffer" loop.

File: lib/system/sysio.nim (hunk 2 above)
Comment (xyz32) [inline lib/system/sysio.nim]: ferror yes, see proc fileError(f: File). 
feof I wanted to use it, and it works in my case, but it breaks "jester", so I left the endOfFile implementation as is.

Comment (dom96) [inline lib/system/sysio.nim]: Oh :(
In what way does it break Jester?

**Comment by xyz32** [comment]
@Araq In my case ftell returns 4098 but fread only reads 7 Bytes, so ``` if readBuffer(file, addr(result[0]), int(len)) != len: ``` will just throw the exception.
Or could we expose readAllBuffer? This method seems to be doing what I am doing in my code for now to work around this issue?

File: lib/system/sysio.nim [hunk 5]
```diff
@@ -147,9 +148,17 @@ proc rawFileSize(file: File): int =
...
+  let bytes = readBuffer(file, addr(result[0]), len)
+  if endOfFile(file):
+    if bytes < len:
+      result.setLen(bytes)
+  elif fileError(file):
     raiseEIO("error while reading from file")
+  else:
+    # We read all the bytes but did not reach the EOF
+    # Try to read it as a buffer
+    result = readAllBuffer(file)
```
Comment (Araq) [inline lib/system/sysio.nim]: How is this correct? You don't append to 'result', you overwrite it with the tail.

Comment (xyz32) [inline lib/system/sysio.nim]: Ye you are right, is not correct. The file stream continues from where I left it.

File: lib/system/sysio.nim (hunk 2 above)
Comment (xyz32) [inline lib/system/sysio.nim]: The static data is no longer sent to the clients (.js, .css etc.). I did not spent time to investigate after seeing this comet: ``` # do not blame me; blame the ANSI C standard this is so brain-damaged ``` inside the endOfFile proc.

File: lib/system.nim [hunk 6]
```diff
@@ -2596,6 +2596,9 @@ when not defined(JS): #and not defined(nimscript):
     proc endOfFile*(f: File): bool {.tags: [], benign.}
       ## Returns true iff `f` is at the end.
 
+    proc fileError*(f: File): bool {.tags: [], benign.}
```
Comment (Araq) [inline lib/system.nim]: I don't want this as an exported system.nim proc.

File: lib/system/sysio.nim [hunk 7]
```diff
@@ -147,9 +148,17 @@ proc rawFileSize(file: File): int =
...
   # We acquire the filesize beforehand and hope it doesn't change.
   # Speeds things up.
-  result = newString(int(len))
-  if readBuffer(file, addr(result[0]), int(len)) != len:
+  result = newString(len)
+  let bytes = readBuffer(file, addr(result[0]), len)
+  if endOfFile(file):
+    if bytes < len:
+      result.setLen(bytes)
+  elif fileError(file):
```
Comment (Araq) [inline lib/system/sysio.nim]: use ferror directly

File: lib/system.nim (hunk 6 above)
Comment (xyz32) [inline lib/system.nim]: Is there a reason for it? I thought if users want to test for the eor they might want to test the error flag as well. Or is it replaced by exceptions?

**Pull request merged**